### PR TITLE
Add Itertools.counts_by.

### PR DESCRIPTION
**Pull request by mmirate**
This is a fairly straightforward extension of the existing "counts" method.

### CODE REVIEW

**Comment by SkiFire13** [comment]
Isn't this exactly the same as calling `Iterator::map` followed by `Itertools::counts`?

**Comment by mmirate** [comment]
Yes.

**Comment by SkiFire13** [comment]
Then I guess you could reuse them to avoid duplicating logic.

File: src/lib.rs [hunk 1]
```diff
@@ -3048,6 +3048,34 @@ pub trait Itertools : Iterator {
...
+    ///     (1, "foo"), (1, "bar"), (1, "baz"),
+    ///     (3, "spam"), (3, "eggs"), (5, "foo")
+    /// ].into_iter().counts_by(|(fst,snd)| fst);
+    /// assert_eq!(counts[&1], 3);
+    /// assert_eq!(counts[&3], 2);
+    /// assert_eq!(counts[&5], 1);
+    /// assert_eq!(counts.get(&0), None);
+    /// ```
+    #[cfg(feature = "use_std")]
+    fn counts_by<K, F>(self, mut f: F) -> HashMap<K, usize>
```
Comment (SkiFire13) [inline src/lib.rs:3071]: `f` doesn't need to be `mut` anymore since now `map` is taking ownership of it

Comment (mmirate) [inline src/lib.rs:3071]:
```suggestion
    fn counts_by<K, F>(self, f: F) -> HashMap<K, usize>
```

Comment (mmirate) [reply src/lib.rs]: Fixed.

File: src/lib.rs [hunk 2]
```diff
@@ -3048,6 +3048,34 @@ pub trait Itertools : Iterator {
...
+    /// # use std::collections::HashMap;
+    /// let counts: HashMap<usize, usize> = vec![
+    ///     (1, "foo"), (1, "bar"), (1, "baz"),
+    ///     (3, "spam"), (3, "eggs"), (5, "foo")
+    /// ].into_iter().counts_by(|(fst,snd)| fst);
+    /// assert_eq!(counts[&1], 3);
+    /// assert_eq!(counts[&3], 2);
+    /// assert_eq!(counts[&5], 1);
+    /// assert_eq!(counts.get(&0), None);
+    /// ```
```
Comment (jswrenn) [inline src/lib.rs:3069]: Let's use an example that feels a bit more real-world; e.g.:
```suggestion
    /// ```
    /// struct Character {
    ///   first_name: &'static str,
    ///   last_name:  &'static str,
    /// }
    /// 
    /// let characters =
    ///     vec![
    ///         Character { first_name: "Amy",   last_name: "Pond"      },
    ///         Character { first_name: "Amy",   last_name: "Wong"      },
    ///         Character { first_name: "Amy",   last_name: "Santiago"  },
    ///         Character { first_name: "James", last_name: "Bond"      },
    ///         Character { first_name: "James", last_name: "Sullivan"  },
    ///         Character { first_name: "James", last_name: "Norington" },
    ///         Character { first_name: "James", last_name: "Kirk"      },
    ///     ];
    /// 
    /// let first_name_frequency = 
    ///     characters
    ///         .into_iter()
    ///         .counts_by(|c| c.first_name);
    ///     
    /// assert_eq!(first_name_frequency["Amy"], 3);
    /// assert_eq!(first_name_frequency["James"], 4);
    /// assert_eq!(first_name_frequency.contains_key("Asha"), false);
    /// ```
```

Comment (jswrenn) [reply src/lib.rs:3069]: This almost makes me wish that `counts` was backed by a `BTreeMap` — then you could naturally iterate over the counted items in order of frequency.

**Review by jswrenn — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (mmirate) [reply src/lib.rs]: Heh. And then some people wonder why we keep chomping at the bit for HKTs. 😄

File: src/lib.rs [hunk 3]
```diff
@@ -3054,18 +3054,31 @@ pub trait Itertools : Iterator {
...
-    /// # use std::collections::HashMap;
-    /// let counts: HashMap<usize, usize> = vec![
-    ///     (1, "foo"), (1, "bar"), (1, "baz"),
-    ///     (3, "spam"), (3, "eggs"), (5, "foo")
-    /// ].into_iter().counts_by(|(fst,snd)| fst);
-    /// assert_eq!(counts[&1], 3);
-    /// assert_eq!(counts[&3], 2);
-    /// assert_eq!(counts[&5], 1);
-    /// assert_eq!(counts.get(&0), None);
+    /// struct Character {
```
Comment (mmirate) [inline src/lib.rs:3058]:
```suggestion
    # use itertools::Itertools;
    /// struct Character {
```

**Review by jswrenn — APPROVED** [review]
Thanks for this!

bors r+

**Pull request merged**